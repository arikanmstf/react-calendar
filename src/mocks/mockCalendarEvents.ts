const mockCalendarEvents = {
  1697760000000: [
    { title: "Mock event", date: new Date(1697760000000), id: "uuid-1" },
    { title: "Second mock event", date: new Date(1697760000000), id: "uuid-2" },
  ],
};

export default mockCalendarEvents;
