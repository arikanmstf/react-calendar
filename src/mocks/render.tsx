import { render as renderRTL } from "@testing-library/react";
import AppContextProvider from "../context/AppContextProvider";
import { ReactElement } from "react";

const render = (ui: ReactElement, { date }: { date?: Date } = {}) =>
  renderRTL(ui, {
    wrapper: ({ children }) => (
      <AppContextProvider date={date || new Date()}>
        {children}
      </AppContextProvider>
    ),
  });

export default render;
