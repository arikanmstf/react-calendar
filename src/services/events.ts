const EVENTS_KEY = "REACT_CALENDAR_APP_EVENTS";

// this typing is here, because probably this depends on backend response.
export type CalendarEvent = {
  date: number; // Date.valueOf
  title: string;
  id: string;
};

function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// GET /api/events
export const getEvents = async (): Promise<Record<number, CalendarEvent[]>> => {
  await sleep(300); // simulate api call.
  const events = window.sessionStorage?.getItem(EVENTS_KEY) || "{}";
  return JSON.parse(events);
};

// POST /api/events
export const postEvent = async (calendarEvent: CalendarEvent) => {
  if (window.sessionStorage) {
    // normally we don't need to fetch here at real api.
    const eventsObject = await getEvents();
    if (eventsObject[calendarEvent.date]) {
      eventsObject[calendarEvent.date].push(calendarEvent);
    } else {
      eventsObject[calendarEvent.date] = [calendarEvent];
    }
    window.sessionStorage.setItem(EVENTS_KEY, JSON.stringify(eventsObject));
  }
};

// DELETE /api/events/${id}
export const deleteEvent = async (calendarEvent: CalendarEvent) => {
  if (window.sessionStorage) {
    const eventsObject = await getEvents();
    if (eventsObject[calendarEvent.date]) {
      eventsObject[calendarEvent.date] = eventsObject[
        calendarEvent.date
      ].filter((event) => event.id !== calendarEvent.id);
      window.sessionStorage.setItem(EVENTS_KEY, JSON.stringify(eventsObject));
    }
  }
};

// delete all events
export const resetEvents = () => {
  window.sessionStorage?.setItem(EVENTS_KEY, "");
};
