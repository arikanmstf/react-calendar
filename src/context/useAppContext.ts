import { useContext } from "react";
import { getEvents } from "../services/events";
import { AppContext, AppDispatchContext } from "./AppContextProvider";

const useAppContext = () => {
  const context = useContext(AppContext);
  const dispatch = useContext(AppDispatchContext);

  const goToNextMonth = () => {
    if (context.monthsToShow.length < 3) {
      return;
    }
    dispatch({ type: "GO_TO_NEXT_MONTH" });
  };
  const goToPreviousMonth = () => {
    if (context.monthsToShow.length < 3) {
      return;
    }
    dispatch({ type: "GO_TO_PREVIOUS_MONTH" });
  };
  const jumpToDate = (date: Date) => {
    console.log("jumpToDatejumpToDate");
    dispatch({ type: "JUMP_TO_DATE", data: date });
  };

  const updateOpenPopoverIndex = (index: string) => {
    dispatch({ type: "UPDATE_OPEN_POPOVER_INDEX", data: index });
  };

  const fetchAndUpdateCalendarEvents = async () => {
    const events = await getEvents();
    dispatch({ type: "UPDATE_CALENDAR_EVENTS", data: events });
  };

  return {
    ...context,
    activeDate: context.monthsToShow[context.monthsToShow.length - 2],
    jumpToDate,
    goToNextMonth,
    goToPreviousMonth,
    updateOpenPopoverIndex,
    fetchAndUpdateCalendarEvents,
  };
};

export default useAppContext;
