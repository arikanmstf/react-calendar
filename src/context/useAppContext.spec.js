import useAppContext from "./useAppContext";
import render from "../mocks/render";
import { screen, waitFor } from "@testing-library/react";
import { useEffect } from "react";
import { getEvents } from "../services/events";
import mockCalendarEvents from "../mocks/mockCalendarEvents";

const LogState = ({ toCall, date, index } = {}) => {
  const {
    goToNextMonth,
    goToPreviousMonth,
    jumpToDate,
    updateOpenPopoverIndex,
    fetchAndUpdateCalendarEvents,
    ...state
  } = useAppContext();

  useEffect(() => {
    switch (toCall) {
      case "goToNextMonth":
        goToNextMonth();
        break;
      case "goToPreviousMonth":
        goToPreviousMonth();
        break;
      case "jumpToDate":
        jumpToDate(date);
        break;
      case "updateOpenPopoverIndex":
        updateOpenPopoverIndex(index);
        break;
      case "fetchAndUpdateCalendarEvents":
        void fetchAndUpdateCalendarEvents();
        break;
      default:
        break;
    }
  }, [toCall]); // eslint-disable-line
  return <pre>{JSON.stringify(state)}</pre>;
};

describe("useAppContext", () => {
  test("timezone should always be UTC", () => {
    expect(new Date().getTimezoneOffset()).toBe(0);
  });

  it("should return state correctly", () => {
    const { container } = render(<LogState />);
    expect(
      screen.getByText(/"activeDate":"2023-10-20T00:00:00.000Z"/),
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should call goToNextMonth correctly", () => {
    const { container } = render(<LogState toCall="goToNextMonth" />);
    expect(
      screen.getByText(/"activeDate":"2023-11-01T00:00:00.000Z"/),
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should call goToPreviousMonth correctly", () => {
    const { container } = render(<LogState toCall="goToPreviousMonth" />);
    expect(
      screen.getByText(/"activeDate":"2023-09-01T00:00:00.000Z"/),
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should call jumpToDate correctly", () => {
    const { container } = render(
      <LogState toCall="jumpToDate" date={new Date("2024-06-29")} />,
    );
    expect(
      screen.getByText(/"activeDate":"2024-06-29T00:00:00.000Z"/),
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should call updateOpenPopoverIndex correctly", () => {
    const { container } = render(
      <LogState toCall="updateOpenPopoverIndex" index="TEST_INDEX" />,
    );
    expect(
      screen.getByText(/"openPopoverIndex":"TEST_INDEX"/),
    ).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it("should call fetchAndUpdateCalendarEvents correctly", async () => {
    getEvents.mockResolvedValue(mockCalendarEvents);
    const { container } = render(
      <LogState toCall="fetchAndUpdateCalendarEvents" />,
    );
    await waitFor(() => {
      expect(
        screen.getByText(/"calendarEvents":{"1697760000000"/),
      ).toBeInTheDocument();
    });
    expect(container).toMatchSnapshot();
  });
});
