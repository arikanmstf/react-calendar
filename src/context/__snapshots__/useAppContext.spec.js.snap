// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useAppContext should call fetchAndUpdateCalendarEvents correctly 1`] = `
<div>
  <pre>
    {"activeDate":"2023-10-20T00:00:00.000Z","monthsToShow":["2023-09-20T00:00:00.000Z","2023-10-20T00:00:00.000Z","2023-11-20T01:00:00.000Z"],"openPopoverIndex":"","calendarEvents":{"1697760000000":[{"title":"Mock event","date":"2023-10-20T00:00:00.000Z","id":"uuid-1"},{"title":"Second mock event","date":"2023-10-20T00:00:00.000Z","id":"uuid-2"}]}}
  </pre>
</div>
`;

exports[`useAppContext should call goToNextMonth correctly 1`] = `
<div>
  <pre>
    {"activeDate":"2023-11-20T01:00:00.000Z","monthsToShow":["2023-10-20T00:00:00.000Z","2023-11-20T01:00:00.000Z","2023-12-20T01:00:00.000Z"],"openPopoverIndex":"","calendarEvents":{}}
  </pre>
</div>
`;

exports[`useAppContext should call goToPreviousMonth correctly 1`] = `
<div>
  <pre>
    {"activeDate":"2023-09-20T00:00:00.000Z","monthsToShow":["2023-08-20T00:00:00.000Z","2023-09-20T00:00:00.000Z","2023-10-20T00:00:00.000Z"],"openPopoverIndex":"","calendarEvents":{}}
  </pre>
</div>
`;

exports[`useAppContext should call jumpToDate correctly 1`] = `
<div>
  <pre>
    {"activeDate":"2024-06-29T00:00:00.000Z","monthsToShow":["2024-05-29T00:00:00.000Z","2024-06-29T00:00:00.000Z","2024-07-29T00:00:00.000Z"],"openPopoverIndex":"","calendarEvents":{}}
  </pre>
</div>
`;

exports[`useAppContext should call updateOpenPopoverIndex correctly 1`] = `
<div>
  <pre>
    {"activeDate":"2023-10-20T00:00:00.000Z","monthsToShow":["2023-09-20T00:00:00.000Z","2023-10-20T00:00:00.000Z","2023-11-20T01:00:00.000Z"],"openPopoverIndex":"TEST_INDEX","calendarEvents":{}}
  </pre>
</div>
`;

exports[`useAppContext should return state correctly 1`] = `
<div>
  <pre>
    {"activeDate":"2023-10-20T00:00:00.000Z","monthsToShow":["2023-09-20T00:00:00.000Z","2023-10-20T00:00:00.000Z","2023-11-20T01:00:00.000Z"],"openPopoverIndex":"","calendarEvents":{}}
  </pre>
</div>
`;
